[alert] Introduce new component

dimitropoulos opened the pull request:
Preview: https://deploy-preview-18702--material-ui.netlify.com/components/alert/.

Closes #18094
Closes #18750

oliviertassinari commented on packages/material-ui-codemod/src/v0.15.0/import-path.js:
```diff
@@ -1,4 +1,5 @@
 const pathConversion = {
+  alert: 'Alert',
```
No needed
```suggestion
```

dimitropoulos replied:
removed in https://github.com/mui-org/material-ui/pull/18702/commits/f217f972843ce74dd02cb12cdac0f286a46b2c68#diff-19bf0e5dcc69095ac3b6192743f39dd0L2

I wasn't sure what this was for - so I added it in an attempt to mitigate things that may go wrong.  What's it for, by the way?

oliviertassinari replied:
This is an old codemod to upgrade from v0.15 to v0.16. I think that the only reason we keep it is that it can help to write future codemod (as source of inspiration).

dimitropoulos replied:
cool.  thanks!

oliviertassinari commented on docs/src/pages.js:
```diff
@@ -113,6 +113,7 @@ const pages = [
         subheader: '/components/lab',
         children: [
           { pathname: '/components/about-the-lab' },
+          { pathname: '/components/alerts' },
```
Singular, house policy

```suggestion
          { pathname: '/components/alert' },
```

dimitropoulos replied:
ahh.  I wasn't sure where to use singular and where to use plural but now that you point it out it makes sense!

r3dm1ke commented:
I have a problem with a test failure in `packages/material-ui/src/styles/createPalette.test.js:681:14` and I cannot figure out why. Can anyone please take a look, maybe I missed something?

oliviertassinari commented:
@r3dm1ke Is the fail related to the new colors in the palette? What if we handle it in another effort?

dimitropoulos commented:
![Screenshot_20191213_100549](https://user-images.githubusercontent.com/15232461/70809838-9fb29b00-1d90-11ea-83c7-9280421e2d00.png)

I only had a few short minutes to spend, but I was able to get some more boilerplate in place.

notes:
1. Regarding the usages of `className={clsx(classes.root, className)}`, I'm not sure how to tell which components need it or not - so I added it everywhere and verified that it does work on all components.  @oliviertassinari your guidance here would be helpful.
1. I have little-to-no idea how we would want to manage the position of elements.  For now, as a hack, I just used absolute positioning - but I think we should come up with something better.  @oliviertassinari your guidance here would be helpful.  I would _love_ to use CSS Grid here ([as I see you would too](https://github.com/mui-org/material-ui/issues/18094#issuecomment-562785773)).  It's absolutely ready for prime time and this would be a very appropriate place to use it.
1. I am not sure which convention to use with `Typography` as it pertains to `AlertTitle` so I just copied from `DialogTitle` and used `<Typography component="h2" variant="h6">`.  @oliviertassinari your guidance would be helpful here.
1. I'm not sure if it's actually appropriate that `AlertDescription` is wrapped with a `Typography` the way `AlertTitle` is.
1. I (had to?) add `@material-ui/icons` to the lab package.  If there was a better/different way to do this please let me know.
  1. Same goes for the way `Typography` was imported in `AlertTitle` and `AlertDescription`.  I'm not entirely sure if it's better to just do `../../../../Typography` or import directly (like I did) from `@material-ui/core`
1. should I have used `@ignore` on the `classes` props?

It's coming along!

oliviertassinari commented:
Sweet :)

1. Using it everywhere sounds great.
2. My main concern with the usage of CSS Grid is around, does the majority of the developers have enough knowledge of it to be able to easily apply customization? I think that it would be great to try flexbox out first. I will see what I can find to help.
3. The usage of the Typography sounds great. I'm wondering about the right element to use, maybe it should be a div?
4. Using the Typography sounds good too.
5. We don't use the icons package in the core to avoid the circular dependency and to keep the installation size minimal. I think that we should apply the same approach to the other icons we use in the core.
6. Import from the public package.
7. In doubt, apply the same approach to the other components, the components should be highly consistent.

dimitropoulos commented:
cool - I removed AlertIcon in https://github.com/mui-org/material-ui/pull/18702/commits/bb5d4c666aaa3ec197d95de5f6fc0f94768422df and instead used a `startIcon` API just like `Button`.  I think it's a great improvement.  It's actually much simpler this way and also has a new feature: the alert will now work just as well without an icon - which is cool.  We [don't need any "hasIcon" prop](https://evergreen.segment.com/components/alert/) or anything.

dimitropoulos commented:
in https://github.com/mui-org/material-ui/pull/18702/commits/1d3d0532ec3afe27ebc353853455cc39ee1a78ce I did a similar treatment for the AlertClose.  Now, the `onClose` functionality for that Close button works, as well as the new feature that the icon is customizable (but defaults to `<Close />` if none is provided).

dimitropoulos commented on docs/src/pages/discover-more/showcase/appList.js:
```diff
@@ -406,7 +406,7 @@ const appList = [
     title: 'Rung - Exceptionality Management',
     description:
-      'Rung alerts you about the exceptionalities of your personal and professional life.',
+      'Rung alert you about the exceptionalities of your personal and professional life.',
```
```suggestion
      'Rung alerts you about the exceptionalities of your personal and professional life.',
```

dimitropoulos replied:
@r3dm1ke this looks like a mistake.

dimitropoulos commented:
in https://github.com/mui-org/material-ui/pull/18702/commits/1b50dd10543e537a2e8922ef379fe8739506d3da I addressed a functionality concern about wrapping in `Typography` for the contents of the alert.  I changed the name to `AlertContent` to align with `CardContent`, `DialogContent`, and `StepContent` and also to signify that it's not just a description, _necessairly_ since (now) someone can pass whatever they want in by specifying a `variant` of `container` on `AlertContent` (the default is `text`).

dimitropoulos commented:
In https://github.com/mui-org/material-ui/pull/18702/commits/82b6c922ed8d55c28eae513d5700be7f45c03251 I was able to spend some time with this after work today and fix some things with the styling.  Here's where it's at currently.  @oliviertassinari is there a design spec anywhere for this (from a material design perspective)?  I wasn't able to find one.

with everything:
![Screenshot_20191213_161819](https://user-images.githubusercontent.com/15232461/70832863-2f257180-1dc4-11ea-84a3-668aea1a0f9d.png)

with no actions:
![Screenshot_20191213_161935](https://user-images.githubusercontent.com/15232461/70832927-5d0ab600-1dc4-11ea-925f-c77833b1a4ca.png)

with no startIcon:
![Screenshot_20191213_162002](https://user-images.githubusercontent.com/15232461/70832946-6e53c280-1dc4-11ea-82e6-a521538897e5.png)

with no closeIcon or Title:
![Screenshot_20191213_162043](https://user-images.githubusercontent.com/15232461/70832962-875c7380-1dc4-11ea-91ad-99963b431125.png)

just the title:
![Screenshot_20191213_162223](https://user-images.githubusercontent.com/15232461/70833047-c2f73d80-1dc4-11ea-9c90-14f2644ac77a.png)

r3dm1ke commented:
I added some initial coloring support:
![image](https://user-images.githubusercontent.com/12103433/70853319-4de63f80-1e7a-11ea-9f06-e27d45e09dd4.png)

There is an issue with coloring the buttons in AlertActions, though. Should we do it, or the user  should provide the colors to the buttons himself?

oliviertassinari commented:
It looks, great guys 😍.

We had an internal discussion with @mbrookes around how we could approach the API on this component and the others. Basically, the root of the concern is on, how close to the underlying DOM node should the components be? Should we have 1. one component with many props or 2. multiple components with a few props.

1. pros:
- More opinionated around how it displays, less question to ask, less room for mistake
- Less code to read and to write
- Each prop gives a strong signal on what the node is about and how it should be displayed, it can make the core logic simpler. 

2. pros:
- allow more flexibility
- simpler to customize and override and it removes a layer of hidden logic. 

Historically, we have been going with 2. as the default or in doubt, and go with 1. when its pros of out weight the pros of 2.

It seems that @dimitropoulos had a first-hand experience of the pros of 1. when moving to ditch the AlertIcon and AlertClose components. It makes sense and goes into @mbrookes that advocates for more 1.
Now, I still think that we should stick to 2. as much as we can. For the text input component, we had to build a TextField component on top of the lower components for simplicity.

In the case of the Alert, I would propose that we keep the current break down of components, I believe most people will use the Alert along. I think that the AlertTitle, AlertDescription, AlertActions are better designed for the long tail of the use cases. What do you guys think?

mbrookes commented:
> I think that the AlertTitle, AlertDescription, AlertActions are better designed for the long tail of the use cases.

Do AltertTitle and AlertDescription add enough value to exist as standalone components, being just a thin wrapper around Typography? Your suggestion to just use Typography directly gives the developer more control of the styling (absent an alert component to follow in the MD spec).

oliviertassinari commented:
Good questions, I don't know.

dimitropoulos commented:
I use material for personal projects and at work because I think that the experience, as with Dialog, of just typing `<DiT` + `Enter` and getting the title to be a superior experience than the experience of having to remember that the dialog title is a `<Typography component="h2" variant="h6">`
![Screenshot_20191216_135655](https://user-images.githubusercontent.com/15232461/70934435-ecee7100-200b-11ea-8e7a-a1ffd5170c39.png).

This is something I quite like about the library.  It's not like, though, I would be upset with a `title` prop, if the opposite were to exist.  I just think you hit the situation where you want to eventually put something else in there and having it just be an inlined component is easier with that use case because what you provide as children more closely match what's actually going to be rendered.  It feels less magical while still being pretty easy to use and work with.

___

Is there a rule against doing both?  To pick on title... why not allow a `title` prop that will literally just inject an `AlertTitle` component for you?  I realize this may be undesirable to some people but just to put it out there..

dimitropoulos commented:
For another example of why there's a time and a place for each approach (and we don't have to go all-or-nothing on either one!) consider the close icon and start icon... it just seemed like reinventing the wheel to make an `AlertIcon` component.  Also... the current positioning model would have to totally change to use absolute position to float the Icon to the left and position it correctly.  I would say that for the icons themselves, it makes sense to just have a `startIcon` prop and a `onClose` prop (paired with an optional `closeIcon` prop).

One of the reasons I changed it is because I referenced the API of a couple of the examples, for example AntD:

![Screenshot_20191216_122344](https://user-images.githubusercontent.com/15232461/70928376-ec031280-1ffe-11ea-8d73-33cddffca7eb.png)

```tsx
    <Alert
      message="Success Tips"
      description="Detailed description and advice about successful copywriting."
      type="success"
      showIcon
    />
    <Alert
      message="Informational Notes"
      description="Additional description and information about copywriting."
      type="info"
      showIcon
    />
    <Alert
      message="Warning"
      description="This is a warning notice about copywriting."
      type="warning"
      showIcon
    />
    <Alert
      message="Error"
      description="This is an error message about copywriting."
      type="error"
      showIcon
    />
```

it's just.. so so _easy_ to use.. You don't have to fuss with any additional imports or anything.  You simply just interact with a few simple props and boom.. it's all ready to go.

I'll also note that I really like the idea of the `showIcon` prop automatically handling which icon actually ends up being rendered with the option to (still) provide a custom icon.

![Screenshot_20191216_154129](https://user-images.githubusercontent.com/15232461/70941474-8b81ce80-201a-11ea-84ed-0366e68f0dac.png)
___


Also.. there are use cases that seem legitimate that go outside the scope of what we would probably enjoy maintaining for `AlertIcon`.  For example, including an image (or svg):

![Capture_d’écran_2019-11-04_à_22 25 24](https://user-images.githubusercontent.com/15232461/70928100-4f407500-1ffe-11ea-86c0-dbf9d8bde045.png)

mbrookes commented:
> I think that the experience, as with Dialog, of just typing <DiT + Enter and getting the title to be a superior experience than the experience of having to remember that the dialog title is a `<Typography component="h2" variant="h6">`

That's a fair point, and nothing to stop someone using Typography instead if they want a different appearance.

> Is there a rule against doing both? To pick on title... why not allow a title prop that will literally just inject an AlertTitle 

Funnily enough, I suggested the same thing when we were discussing it. The one downside I can think of is that when you give someone choice about how to do something, it leads to cognitive overhead in understanding the tradeoffs (if any), in order to decide which to use – Grid or Box for layout? Modal, or Popper, or Popover?

Other than that, no strong objection - and we could make it clear in the docs that they are alternate but functionally identical APIs. @oliviertassinari may have other opinions?

> it just seemed like reinventing the wheel to make an AlertIcon component

Not disagreeing, but Stepper has a StepIcon component that is used internally in StepLabel (or can be overridden with the `StepIconComponent` prop). StepIcon is also part of the public API (although we don't have any examples of how you would use it). Stepper might be slightly over-engineered though, but I didn't dig into it with a critical eye – it's a complex component!

SpeedDial also has the SpeedDialIcon component, but it has to be supplied to the SpeedDial via the  `icon` prop, rather than being used internally. This could possibly be reworked to follow convention if the component gets moved to core.

> I'll also note that I really like the idea of the showIcon prop automatically handling which icon actually ends up being rendered

We could possibly import the default icons (as internal icons), and allow the `icon` prop to optionally accept either a string ( 'success' | 'info' | 'warning' | 'error' ) or an actual icon component.

r3dm1ke commented:
Just to chip in my 5 cents: I am in favour of having `AlertTitle`, `AlertDescription` and even `AlertIcon`. If we move everything into one component and pass content via props, the component would be way too big, logic way too complex and it will be harder to unit test. By having dedicated components for every aspect of the alert, we let user decide what exactly does he need and he can later write his own component that will pass props to `Alert[Title, Description, etc.]`. Alternatively, we could keep all these components and make an `AlertBase` component, very much like in `TextField`/`InputBase`.

Also, I have thought about passing the color to the buttons in `AlertActions` and it does not look viable. I think it is better to let user color the buttons himself, because they most likely will have different color based on context (e.g. delete button will most likely be `error` color, send button may be `success` or `primary`, etc).

mbrookes commented:
Given that only one of the eleven Alert benchmark examples in the associated issue has buttons, I was going to question whether Alert should have them, as they seem to belong more in the domain of the Dialog. However, when checking the MD dialog guidelines for a suitable citation as to why, I noticed a reference to the banner component:

https://material.io/components/banners/banners.html

This seems to be as close as the MD spec has to an Alert, but without the fairly standard colors and icons seen in most other implementations. 

How should we apply this here?

dimitropoulos commented:
That's true.  Since making `AlertContent` have a `variant="container"` mode, if the user really wants buttons they can add them (and style them) themselves.

![Screenshot_20191217_054852](https://user-images.githubusercontent.com/15232461/70989050-ef46de80-2090-11ea-8895-35ede97bbde3.png)

I'll just highlight what AntD does, which is sorta a middleground approach.  They don't have an explicit action button section, but you can optionally provide text to be used instead of a close button.  That makes sense, I think, because (if you think about it) most of the time what will semantically happen when you click an action button is somewhat similar to a dismiss or close button.

___

re the lack of a spec for this component:

@oliviertassinari @mbrookes according to https://youtu.be/E23IWqE4nac?t=351 the material design team is supposed to be accessible to us in some way to answer exactly-this-kind-of-question.  Is that the case? They seem to be calling out this project in particular as one they will be helping.  If so can we maybe take that avenue on getting questions answered about the buttons (and, for that matter, overall styling questions like @r3dm1ke's questions about whether there should be elevation/paper/dropshadow on the component or not)?

r3dm1ke commented:
> https://material.io/components/banners/banners.html
> 
> This seems to be as close as the MD spec has to an Alert, but without the fairly standard colors and icons seen in most other implementations.
> 
> How should we apply this here?

In the [discussion](https://github.com/mui-org/material-ui/issues/18094#issuecomment-564677649) for the issue related to this component, there was a suggestion to have a `variant` prop, which way be `'text' | 'outlined' | 'filled' | 'banner'`.

mbrookes commented:
@r3dm1ke Ah, thanks, missed that. What is `text` variant in this instance? (No background color or outline?)

To answer the other question you asked there, yes, this should be a controlled component (`open` prop).

r3dm1ke commented:
@mbrookes the `text` variant should be something like the _secondary alert_ from [here](https://dribbble.com/shots/7874860-Alerts-UI-design/attachments/486141?mode=media), as suggested by @oliviertassinari

mbrookes commented:
Okay, thanks. Not totally convinced by the prop name, but we can come back to it.

oliviertassinari reviewed (commented):
> the text variant should be something like the secondary alert from here

@mbrookes The name is inspired by https://vuetifyjs.com/en/components/alerts#usage. I'm not fond of the name either. Still, I think that it could make a good default variant. How could be better name it?

Regarding the "simple" demo, what about doing something like this instead? Most of the time, I don't think that we need to pull in the title or the actions:
```jsx
<Alert color="primary">Sorry, results couldn't be loaded</Alert>
```

I'm happy to see the effort progress :)!

oliviertassinari commented:
> > I'll also note that I really like the idea of the showIcon prop automatically handling which icon actually ends up being rendered

> We could possibly import the default icons (as internal icons), and allow the icon prop to optionally accept either a string ( 'success' | 'info' | 'warning' | 'error' ) or an actual icon component.

This sounds great!

Regarding the API, I think that we can always change it later on before we move the component in the core. I personally think that we have an acceptable tradeoff, the most common use case requires a single component, it's not bloated by the long tail of rendering a title or actions, so 👍 from my part.

merceyz commented on packages/material-ui-lab/src/Alert/Alert.d.ts:
```diff
@@ -0,0 +1,46 @@
+
+declare const Alert: React.ComponentType<AlertProps>;
+
+export default Alert;
```
```suggestion
export default function Alert(props: AlertProps): JSX.Element;
```

dimitropoulos replied:
This is a syntax error, is it not?

merceyz commented on packages/material-ui-lab/src/Alert/Alert.d.ts:
```diff
@@ -0,0 +1,46 @@
+export type AlertClassKey = 'root';
+
+declare const Alert: React.ComponentType<AlertProps>;
+
```
```suggestion
```

merceyz commented on packages/material-ui-lab/src/Alert/Alert.d.ts:
```diff
@@ -0,0 +1,46 @@
+
+export type AlertClassKey = 'root';
+
+declare const Alert: React.ComponentType<AlertProps>;
```
```suggestion
```

merceyz reviewed (commented):
`typescript-to-proptypes` should be able to handle this component.

r3dm1ke commented on packages/material-ui-lab/src/Alert/Alert.d.ts:
```diff
@@ -0,0 +1,46 @@
+  /**
+   * If truthy, the Alert is visible
+   */
+  open?: boolean;
```
@mbrookes ?

dimitropoulos replied:
I strongly prefer:

```tsx
{open && <Alert>...</Alert>}
```

Personally, because it reduces the number of components in the tree.

dimitropoulos replied:
removed in: https://github.com/mui-org/material-ui/pull/18702/commits/a67b7f35030db3a5dcbd86f817fa853c1c42c00f

dimitropoulos commented on docs/src/pages/components/alert/alert.md:
```diff
@@ -0,0 +1,20 @@
+
+# Alerts
+
+<p class="description">Alert generates a small Alert to the top-right of its child(ren).</p>
```
I originally copied from `Badge` when I was trying to figure out how to create a component and Badge has the description:

> <p class="description">Badge generates a small badge to the top-right of its child(ren).</p>

dimitropoulos commented on packages/material-ui-lab/src/Alert/Alert.d.ts:
```diff
@@ -0,0 +1,46 @@
+   * Element placed before the children.
+   */
+  closeIcon?: React.ReactNode;
+
```
for what it's worth, I personally STRONGLY STRONGLY prefer having an extra line in this context (before the JSDoc starts), but I removed all the extra lines in b9ea09d49d920a2efe37af8918ac6a6e84f670d8 to keep things consistent with the rest of the project.

oliviertassinari replied:
Thank you for putting your personal preference aside for the consistency of the codebase 🙏!

dimitropoulos commented on packages/material-ui-lab/src/Alert/Alert.d.ts:
```diff
@@ -0,0 +1,46 @@
+  /*
+   * Main color for the Alert, picked from theme palette
+   */
+  color?: 'primary' | 'secondary' | 'default' | 'error';
```
yep.  done.

dimitropoulos commented:
The good news:

![Screenshot_20191219_232618](https://user-images.githubusercontent.com/15232461/71230267-f7806300-22b6-11ea-89f5-c98358fad627.png)

- I was able to implement the default icons thing! yay!  You can still provide a custom icon if you want.

The bad news:

I think we are blocked by https://github.com/mui-org/material-ui/issues/13875 for one thing because I just picked colors somewhat randomly, and for another thing because it would be really really really nice if you could pass the `type` (i.e. `severity` or whatever) to the `Buttons` in the `AlertActions` component.

oliviertassinari commented on docs/src/pages/components/alert/SimpleAlert.js:
```diff
@@ -0,0 +1,69 @@
+        <AlertTitle>Error!</AlertTitle>
+        <AlertContent>Your request for the ultimate cup of coffee has been denied!</AlertContent>
+      </Alert>
+
```
```suggestion
```

oliviertassinari commented on docs/src/pages/components/alert/SimpleAlert.js:
```diff
@@ -0,0 +1,69 @@
+  },
+})(Box);
+
+const Alert = withStyles({
```
I think that we can merge these two withStyles into a single statement (to make it easier to copy and paste in a codebase). Something like this:
```jsx
import { makeStyles, Theme, createStyles } from '@material-ui/core/styles';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      '& > * + *': {
        marginTop: theme.spacing(2),
      },
    },
  }),
);
```
/docs/src/pages/components/progress/CircularDeterminate.tsx

oliviertassinari commented on packages/material-ui-lab/src/Alert/Alert.js:
```diff
@@ -0,0 +1,153 @@
+    position: 'relative',
+    display: 'flex',
+    alignItems: 'center',
+    padding: '1em 0',
```
We haven't been using em unit in the codebase much (if none at all), what about using a multiple of 8, in pixels? (the underlying grid uses 4px).

oliviertassinari commented on packages/material-ui-lab/src/Alert/Alert.js:
```diff
@@ -0,0 +1,153 @@
+  },
+});
+
+const getDefaultIcon = (type) => {
```
What about using an object map? We could also imagine exposing an `iconMapping` prop like we have with the Typography and variantMapping.

oliviertassinari commented on packages/material-ui-lab/src/AlertActions/AlertActions.js:
```diff
@@ -0,0 +1,57 @@
+    justifyContent: 'flex-start',
+    flex: '0 0 auto',
+  },
+  /** Styles applied to the root element if `disableSpacing={false}`. */
```
```suggestion
  /* Styles applied to the root element if `disableSpacing={false}`. */
```

oliviertassinari commented on packages/material-ui-lab/src/AlertTitle/AlertTitle.js:
```diff
@@ -0,0 +1,44 @@
+  const { classes, children, className } = props;
+  return (
+    <div className={clsx(classes.root, className)} ref={ref}>
+      <Typography component="h2" variant="h6">
```
I think that we should merge these two-element, it should render a single DOM node.

dimitropoulos committed:
[Alert] implements Alert component

r3dm1ke commented:
Since the additional colors are now merged into master, I have hooked up the component to use them. And I found out that the text is not contrastive enough as it is (especially at warning):
![image](https://user-images.githubusercontent.com/12103433/71381819-3823f900-25a3-11ea-9e1b-933d2e764522.png)
This may be fixed by using the main color as the background, and `contrastText` color for the text color:
![image](https://user-images.githubusercontent.com/12103433/71381796-1fb3de80-25a3-11ea-8eba-d84719b3e6be.png)

What do you think?

r3dm1ke committed:
Fixing issues from code review

Fixing prettier issues


Changing units from em to px


Picking up additional colors from palette


Refactored default icon lookup


Refactored AlertTitle into rendering 1 DOM node

dimitropoulos commented:
wow great stuff!!! I really was hoping to get to that this weekend but it didn't end up happening.

They're both kinda tough for their own reasons.. huh.  @oliviertassinari @r3dm1ke did you see what I said above about asking someone from the actual material-design team?  They claim they want to help, and I would tend to believe them saying that.  https://github.com/mui-org/material-ui/pull/18702#issuecomment-566491127

oliviertassinari committed:
push one step further

oliviertassinari commented:
Nice :). I'm pushing it one step further.

mbrookes commented:
Ant solves it by using dark text on the light background:

![image](https://user-images.githubusercontent.com/357702/71413740-7f65c480-264b-11ea-8c3a-4f01deac777c.png)

mbrookes commented:
For the customized snackbars example, `orange[500]` for `warning` gives better contrast than `amber[500]`:

![image](https://user-images.githubusercontent.com/357702/71415100-0d44ae00-2652-11ea-8264-189510c04a4a.png)


![image](https://user-images.githubusercontent.com/357702/71415031-b9d26000-2651-11ea-9c4f-8414a4d0e15e.png)

oliviertassinari commented:
I have tried to push further the effort. There are two points I'm wondering about:
- I have removed the `onClose` prop, but I'm wondering if we shouldn't add it back. On one hand, this enables tree shaking, on the other, it requires more imports and manual translation of the close action.
- The variant and color props are marked as `@ignored` during the TypeScript conversion. I don't understand why.

dimitropoulos commented:
> I have removed the onClose prop, but I'm wondering if we shouldn't add it back. On one hand, this enables tree shaking, on the other, it requires more imports and manual translation of the close action.

It's not quite ready yet, but I'm inclined to show you a sneak-peek of a little project I've been working on as a result of this PR: https://github.com/dimitropoulos/react-ui-roundup#Alert (or you can (pretty much) the same thing with a few more features at https://react-ui-roundup.dimitrimitropoulos.com).  If you look at all of the Alert components, you'll see that `onClose` functionality is pretty standard equipment.  It'd be great to keep it in there.  If there are build size issues (i.e related to the tree shaking) perhaps we can loop back on this in another iteration.  It is in the lab, after all.

> The variant and color props are marked as @ignored during the TypeScript conversion. I don't understand why.

This is almost definitely a mistake.  If marked as such by me, it was done so in ignorance.

oliviertassinari commented:
> If you look at all of the Alert components, you'll see that onClose functionality is pretty standard equipment.

@dimitropoulos Thanks for the benchmark! Yeah, I like the idea. I think that @mbrookes would support the prop too.

> This is almost definitely a mistake. If marked as such by me, it was done so in ignorance.

I suspect an issue with the TypeScript to prop types generation script.

merceyz commented on packages/material-ui-lab/src/Alert/Alert.d.ts:
```diff
@@ -0,0 +1,55 @@
+   * The action to display.
+   */
+  action?: React.ReactNode;
+  /*
```
```suggestion
  /**
```

oliviertassinari replied:
🎄 🎉

merceyz commented on packages/material-ui-lab/src/Alert/Alert.d.ts:
```diff
@@ -0,0 +1,55 @@
+   * Alternatively, you can use the `icon` prop.
+   */
+  iconMapping?: Partial<Record<Color, React.ReactNode>>;
+  /*
```
```suggestion
  /**
```

merceyz reviewed (changes requested):
> The variant and color props are marked as @ignored during the TypeScript conversion. I don't understand why.

@oliviertassinari They're marked as ignored because they lack a JSDoc comment, it's currently just a multiline comment

https://github.com/mui-org/material-ui/blob/862f22413f0c462a6b24bdaa4aa84169cf4cb2a7/scripts/generateProptypes.ts#L52-L54

merceyz committed:
help with api docs

oliviertassinari committed:
Add onClose prop back

oliviertassinari reviewed (approved):
I have fixed the two previous issues, with the help of @merceyz 🤗.
If you guys see something that we should improve, please do raise it :).
Notice that there are a couple of simplifications since we first iterated on the component.
I have removed the AlertActions component, I would propose that we wait for user requests.
As a future improvement, we could probably add a customization demo for the alert, one that implements a bordered variant, e.g. https://evergreen.segment.com/components/alert or https://sancho-ui.com/components/alert/. Or even, we could demonstrate how to build https://chakra-ui.com/alert#composition (with 3 CSS properties).

Great work guys, please review.

oliviertassinari commented:
@dimitropoulos To answer your question regarding getting feedback from the Material Design team. I could find the following, I think that we are on the right track with the default `text` variant (that matt might want to rename):

![Capture d’écran 2019-12-25 à 19 07 15](https://user-images.githubusercontent.com/3165635/71449300-ef5a7480-2749-11ea-8a96-2c299bec34dc.png)
https://admin.google.com/ac/accountchooser?continue=https://admin.google.com/ac/devices/list?category%3Dmobile

![Capture d’écran 2019-12-25 à 19 10 12](https://user-images.githubusercontent.com/3165635/71449315-334d7980-274a-11ea-91a3-d48e6859db33.png)
https://console.firebase.google.com/project/react-swipeable-views-395e4/hosting/main

![Capture d’écran 2019-12-25 à 19 12 58](https://user-images.githubusercontent.com/3165635/71449344-95a67a00-274a-11ea-8c70-0c904414e808.png)
https://developers.google.com/maps/documentation/roads/intro

We have been in contact with the Material Design team (personally did 2 calls and one on physical meeting so far). However, I wouldn't expect to receive specific guidance on this matter. At least, I would be positively surprised if we do.

dimitropoulos commented on packages/material-ui-lab/src/internal/svg-icons/ErrorOutlined.js:
```diff
@@ -0,0 +1,10 @@
+ */
+export default createSvgIcon(
+  <path d="M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41, 20 12,20M12,2C6.47,2 2,6.47 2,12C2,17.53 6.47,22 12,22C17.53,22 22,17.53 22,12C22,6.47 17.53, 2 12,2M14.59,8L12,10.59L9.41,8L8,9.41L10.59,12L8,14.59L9.41,16L12,13.41L14.59,16L16, 14.59L13.41,12L16,9.41L14.59,8Z" />,
+  'ErrorOutlined',
```
![Screenshot_20191226_070016](https://user-images.githubusercontent.com/15232461/71475244-61b07080-27ad-11ea-9ca1-b0bed5cc069f.png)

![Screenshot_20191226_070032](https://user-images.githubusercontent.com/15232461/71475251-6bd26f00-27ad-11ea-9383-23153115c920.png)

I guess this just kinda "is what it is" since this is up to material-design more than anything but it'd be cool if there was a better icon to use here that didn't look like a close icon.  I'm imaging this being used in context, where you'd rarely have more than one alert.

oliviertassinari replied:
The current icons are inspired by https://iamhosseindhv.com/notistack/demos from @iamhosseindhv. Would "ErrorOutiline" be a better fit?

![Capture d’écran 2019-12-26 à 13 12 20](https://user-images.githubusercontent.com/3165635/71475624-6000a000-27e1-11ea-8ae7-4e1deec3ca00.png)

```jsx
import ErrorOutlineIcon from '@material-ui/icons/ErrorOutline';
```

oliviertassinari replied:
The current icons are inspired by https://iamhosseindhv.com/notistack/demos from @iamhosseindhv. Would "ErrorOutiline" be a better fit?

![Capture d’écran 2019-12-26 à 13 12 20](https://user-images.githubusercontent.com/3165635/71475624-6000a000-27e1-11ea-8ae7-4e1deec3ca00.png)

```jsx
import ErrorOutlineIcon from '@material-ui/icons/ErrorOutline';
```

oliviertassinari replied:
![Capture d’écran 2019-12-26 à 14 49 24](https://user-images.githubusercontent.com/3165635/71478260-f4253400-27ee-11ea-8ebd-686697da6cfd.png)

@dimitropoulos I think that it looks better. Do you want to replace it?

dimitropoulos replied:
cool: changed in https://github.com/mui-org/material-ui/pull/18702/commits/198e507da5d3f1e3470e2cac6e976b5a7d2ef93e

dimitropoulos commented on packages/material-ui-lab/src/Alert/Alert.d.ts:
```diff
@@ -0,0 +1,68 @@
+  /**
+   * Main color for the Alert, picked from theme palette.
+   */
+  color?: Color;
```
This is my only major gripe.  I really think we should consider naming this `type`.  For one reason, because we don't actually know what the "color" is that the user has actually set.  It's perfectly possible that two alerts could have the same color.  Furthermore... `export type Color = 'success' | 'info' | 'warning' | 'error';` is in itself a code smell since those things listed are not themselves colors.

For another reason, of the 18 benchmark examples:
- 7 of them use `type`
- 4 of them use `variant` (note: we are already using `variant` for another purpose)
- 2 of them use `intent`
- 1 each use `role`, `color`, `severity`, `kind`, `status`

dimitropoulos replied:
ok.. why not both then?  I guess I was moreso trying to point out that this is controlling more than just the color (i.e. the icon) and you run into problematic questions like:

> what icon should be used for `color="secondary"`

whereas if you keep them separate you don't have this problem.  more than that.. I'm not convinced `primary` and `secondary` actually even make sense in the content of an alert's _type_.  Color.. ok sure.  but the type?  What's a `primary` alert? an error?

If there is an issue for dynamic prop support from the theme please link it here.

oliviertassinari replied:
I don't have a strong point of view on this matter. I have personally built a custom component to solve this problem in the past and called the prop: `variant` and like the simplicity of a single prop even if it mean losing some of the subtilities.

However, the mixed approach: type (that determines the icon and the default color) + color (that can override the color from type) + icon (that can override the icon from type) + iconMapping (that can extend the scope of the type supported) props could work.

@mbrookes What do you think?

dimitropoulos commented on packages/material-ui-lab/src/Alert/Alert.d.ts:
```diff
@@ -0,0 +1,68 @@
+   */
+  icon?: React.ReactNode | false;
+  /**
+   * The role attribute of the element.
```
This description doesn't really say enough.  I'd suggest something to add but I don't, myself, really understand what it does.  It's an accessibility thing or for the snackbar?

oliviertassinari replied:
It's somewhat related to the Snackbar. I propose we link this page: `/components/alert/#accessibility`.

dimitropoulos commented on packages/material-ui-lab/src/Alert/Alert.d.ts:
```diff
@@ -0,0 +1,68 @@
+
+export interface AlertProps extends StandardProps<PaperProps, AlertClassKey, 'variant'> {
+  /**
+   * The action to display.
```
It may be worth clarifying that this (from a DOM perspective) in the position of the close button.  I'm not sure how to word that though.  No biggie either way.

oliviertassinari replied:
It's copied and paste from `SnackbarContent.actions`'s description. Happy to improve both at the same time. Maybe?

```
The action to display. It renders after the message, at the end of the alert.
```

dimitropoulos replied:
sounds good to me.

dimitropoulos replied:
updated in https://github.com/mui-org/material-ui/pull/18702/commits/ffd00685bda70b66868f6f911be8683235c1b132

dimitropoulos commented on packages/material-ui-lab/src/Alert/Alert.js:
```diff
@@ -0,0 +1,269 @@
+          {icon || iconMapping[color] || defaultIconMapping[color]}
+        </div>
+      ) : null}
+      <div className={classes.message}>{children}</div>
```
Just verifying: there's no reason why `children` needs to be wrapped in `Typography` here, correct?

oliviertassinari replied:
No need to wrap children in Typography, correct.

dimitropoulos commented on packages/material-ui-lab/src/Alert/Alert.d.ts:
```diff
@@ -0,0 +1,68 @@
+   */
+  action?: React.ReactNode;
+  /**
+   * Override the default text for the *close popup* icon button.
```
```suggestion
   * Override the default tooltip text for the *close popup* icon button, if used.
```

dimitropoulos committed:
changes error icon to `ErrorOutline` from `ErrorOutlined`

dimitropoulos committed:
updates description for `action` prop

oliviertassinari commented on packages/material-ui-lab/src/internal/svg-icons/ErrorOutline.js:
```diff
@@ -0,0 +1,10 @@
+ */
+export default createSvgIcon(
+  <path d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z" />,
+  'ErrorOutlined',
```
```suggestion
  'ErrorOutline',
```

oliviertassinari commented on packages/material-ui/src/SnackbarContent/SnackbarContent.js:
```diff
@@ -61,7 +61,7 @@ const SnackbarContent = React.forwardRef(function SnackbarContent(props, ref) {
 SnackbarContent.propTypes = {
   /**
-   * The action to display.
+   * The action to display. It renders after the message, at the end of the alert.
```
?
```suggestion
   * The action to display. It renders after the message, at the end of the snackbar.
```

dimitropoulos replied:
yikes!

dimitropoulos committed:
Update packages/material-ui-lab/src/internal/svg-icons/ErrorOutline.js

Co-Authored-By: Olivier Tassinari <olivier.tassinari@gmail.com>

dimitropoulos committed:
Update packages/material-ui/src/SnackbarContent/SnackbarContent.js

Co-Authored-By: Olivier Tassinari <olivier.tassinari@gmail.com>

mbrookes commented:
My first thought was that, if in doubt, we should go with the simplest API first (`type`), and add the `color` and/or `icon` props later based on demand; however those particular props are pretty ubiquitous, so I'm not so sure it makes sense to omit them.

My only other niggle is with the prop name `type`. We've avoided it in the past as an overloaded term. `severity` is the best alternative I can think of.

dimitropoulos commented:
for what it's worth: I, too, feel that `severity` is the next-best choice.  I can understand the desire to stay away from `type`.

oliviertassinari commented:
Does it mean we have found a common understanding under this approach:

type (that determines the icon and the default color) + color (that can override the color from type) + icon (that can override the icon from type) + iconMapping (that can extend the scope of the type supported) props could work.

?

oliviertassinari committed:
Matt review

mbrookes committed:
[Alert] Fix demo message dash spacing

mbrookes commented on packages/material-ui-lab/src/Alert/Alert.js:
```diff
@@ -0,0 +1,278 @@
+   */
+  className: PropTypes.string,
+  /**
+   * Override the default text for the *close popup* icon button.
```
```suggestion
   * Override the default text for the *close* icon button's aria-label.
```

```suggestion
   * Override the default aria-label for the *close* icon button.
```

mbrookes commented on packages/material-ui-lab/src/Alert/Alert.js:
```diff
@@ -0,0 +1,278 @@
+  icon: PropTypes.node,
+  /**
+   * The component maps the color prop to a range of different icons.
+   * For instance, success to `<SuccessOutlined>`.
```
```suggestion
   * for instance success to `<SuccessOutlined>`.
```

mbrookes commented on packages/material-ui-lab/src/Alert/Alert.js:
```diff
@@ -0,0 +1,278 @@
+   */
+  role: PropTypes.string,
+  /**
+   * The severity for the alert.
```
```suggestion
   * The severity of the alert. This defines the color and icon used.
```

mbrookes commented on packages/material-ui-lab/src/Alert/Alert.js:
```diff
@@ -0,0 +1,278 @@
+  }),
+  /**
+   * Callback fired when the component requests to be closed.
+   * When provided and no action prop is set, a close icon is displayed.
```
```suggestion
   * When provided and no action prop is set, a close icon button is displayed that triggers the callback when clicked.
```

mbrookes commented on packages/material-ui-lab/src/Alert/Alert.js:
```diff
@@ -0,0 +1,278 @@
+  /**
+   * The component maps the color prop to a range of different icons.
+   * For instance, success to `<SuccessOutlined>`.
+   * If you wish to change that mapping, you can provide your own.
```
```suggestion
   * If you wish to change this mapping, you can provide your own.
```

mbrookes commented on packages/material-ui-lab/src/Alert/Alert.js:
```diff
@@ -0,0 +1,278 @@
+   * The component maps the color prop to a range of different icons.
+   * For instance, success to `<SuccessOutlined>`.
+   * If you wish to change that mapping, you can provide your own.
+   * Alternatively, you can use the `icon` prop.
```
```suggestion
   * Alternatively, you can use the `icon` prop to override the icon displayed.
```

mbrookes commented on packages/material-ui-lab/src/Alert/Alert.js:
```diff
@@ -0,0 +1,278 @@
+    >
+      {icon !== false ? (
+        <div className={classes.icon}>
+          {icon || iconMapping[severity] || defaultIconMapping[severity]}
```
```suggestion
          {icon || iconMapping[severity]}
```

oliviertassinari replied:
This logic is important for fallback.

mbrookes replied:
What am I missing? https://github.com/mui-org/material-ui/pull/18702/files/088cf5d8390fcf194f8cdc14248224f625444e59#diff-ba59df5f895c0b27cff62d4e541ac689R164

mbrookes commented on packages/material-ui-lab/src/Alert/Alert.js:
```diff
@@ -0,0 +1,278 @@
+          <IconButton
+            size="small"
+            aria-label={closeText}
+            title={closeText}
```
```suggestion
```

oliviertassinari replied:
Why?

mbrookes replied:
If you want a tooltip use the tooltip component.

oliviertassinari replied:
It made me think of this change: https://github.com/mui-org/material-ui/pull/18283/files#r344488885.

oliviertassinari replied:
Same approach in https://github.com/mui-org/material-ui/blob/a07f18029a5532de310f234392c18664817f0dce/packages/material-ui/src/TablePagination/TablePagination.js#L156-L157

oliviertassinari replied:
The Tooltip component can be overkill for some use cases.

mbrookes replied:
The [tooltip guidelines](https://material.io/components/tooltips/tooltips.html) don't give any advice on using the browser tooltip. When should one be prioritised over the other?

oliviertassinari replied:
From my perspective, with an on-side example:

Tooltip pros:
- control the display delay, displays instantly,
- looks nice
![Capture d’écran 2020-01-02 à 00 18 46](https://user-images.githubusercontent.com/3165635/71647334-995f7100-2cf5-11ea-818a-0d3e44791fe2.png)

title pros:
- no bundle size impact
- displays after 2s, nice for non important labels.
![Capture d’écran 2020-01-02 à 00 19 43](https://user-images.githubusercontent.com/3165635/71647335-995f7100-2cf5-11ea-9a08-b6ba5e57cbd8.png)

oliviertassinari commented on docs/src/pages/components/alert/alert.md:
```diff
@@ -0,0 +1,64 @@
+
+## Toast
+
+You can use the Snackbar to [display a toast](/components/snackbars/#customized-snackbars) with the Alert.
```
As far as I know, a toast === a snackbar in the standard language.

mbrookes replied:
In that case the description is tautological.

That table reminds me though - shouldn't this component support the [Banner](https://material.io/components/banners/banners.html) use case, possibly as the default?

oliviertassinari commented on packages/material-ui-lab/src/Alert/Alert.js:
```diff
@@ -0,0 +1,278 @@
+  /**
+   * The variant to use.
+   */
+  variant: PropTypes.oneOf(['filled', 'outlined', 'text']),
```
👍

mbrookes committed:
review

dimitropoulos commented:
looking like we're pretty close to a minimally viable component for the lab!  is there anything else here we have in mind to do? @oliviertassinari @r3dm1ke @mbrookes?

oliviertassinari commented:
I think that we are good to go. @mbrookes has uncovered an interesting topic on title vs Tooltip usage, but it's leaking from the Alert to the TablePagination and Autocomplete. I don't think that it should block this pull request.

oliviertassinari commented:
Thank you everybody for your time!

We will likely have a couple of follow-up pull requests, one step at the time :).

dimitropoulos commented:
woooohoooo!  awesome!  I'm more than happy to help with any followup work so please @oliviertassinari feel free to ping/assign me.

oliviertassinari commented:
@dimitropoulos Thanks! GitHub has this concept of code owner: https://help.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners. It's something that we have never used yet. However, I think that it would be interesting to experiment. Would you be interested in being the code owner of the Alert component? @mbrookes What do you think

mbrookes commented:
I didn't know about that feature. Worth trying!

oliviertassinari commented:
Awesome, let's experiment! I will handle this in my next batch of small changes (probably Sunday).
